### DISCUSSION TITLE
Online Decoding of Speech Input to Microphone

### DISCUSSION THREAD

**Discussion by Grottoh** [flashlight/flashlight#601] [Q&A] [answered]
Hello,

I’ve recently started looking into wav2letter (i.e. Flashlight’s ASR application) and so far I really like it. However I have had some issues, in particular I am currently having issues with the online decoding of speech. To avoid confusion when I say I want to be able to do online decoding of speech (I sometimes feel it does not mean what I think it means): I mean I want wav2letter to be able to process my speech as I’m speaking into my microphone. From what I gather this should be possible, however though I have dedicated no small amount to figuring it out, I am still not clear on how this should be done. The most important things that I’ve been able to find on this topic are:

- https://github.com/flashlight/wav2letter/commit/eda528dd38a509e35cb96ddca492095ec88da1cc#diff-4c1ef2efc13c786158e726976efa154e1a15140146c4e7bf6045577cb8242299
        - Basically this refers to an update of the file that is now <flashlight/flashlight/lib/text/decoder/Decoder.h>. Comments are included in this file which indicate the presence decoder support in an online manner. Mainly it is unclear to me how to directly use microphone input as the audio input stream (though possibly this is due to my current lack of experience with C++).

- https://github.com/flashlight/flashlight/discussions/484
        - The question asked here appears to be related to mine. Vineelpratap provides some resources in response, but these concern the ‘old, pre-consolidation version of wav2letter’, and they do not appear to be applicable to the new consolidated wav2letter (i.e. Flashlight’s ASR application).

My question then is as follows:

- Is what I want (to have wav2letter process speech as it is being spoken into a microphone) possible at present (with the new consolidated version of wav2letter)?
        - If so, should the above links be enough for me to figure out how to do this? Or have I missed something?

I would very much appreciate to be pointed in the right direction on this, or else to let me know that this is a dead end (or requires me to implement it myself). Also if it is obvious that I’ve missed something or made some wrong assumptions that would be great to know.

Thanks

**Comment by gevorgyana** [comment]
Hello, here is what I want to say. It should be possible, at least because the technology they use in flashlight/asr (CTC decoding) allows it, (Seq2Seq, the other strategy they use, however, doesn't), as they mention in the root README.md. Let me try to figure this out and share my findings. I am trying to do the same thing, by the way.

**Reply by Grottoh** [reply]
Hey, thanks for the update, I'll be curious to hear what you find. I won't have time the next few days to figure this out myself, but once I do learn more I'll of course share it here as well.

**Reply by gevorgyana** [reply]
Well, the same thing. I will have to pause for a little bit, but whenever I find info, I share it here. Hope we will get it working :)

**Reply by gevorgyana** [reply]
I asked a similar question here https://github.com/theblackcat102/edgedict/issues/30

**Reply by gevorgyana** [reply]
For online recognition, given we are talking about CTC, we have the following picture. The model is encoder-decoder, the encoder is trained to output tokens from the alphabet, extended by a special null token. The model listens to each frame that comes into it, independently, and outputs a probability distribution for each possible symbol from the extended alphabet. The model is trained to minimize the log-loss, or, which is the same, categorical cross-entropy. Usually, we transcribe the whole audio file. But let's think about online ASR. What is 'online'? If you think about it, we have certain constraints: we can't predict anything just by looking at a few consecutive frames, because that would be too little to even recognize a letter utterance. Thus, we can clearly differentiate between words when they are spoken with pauses in between them. This way, we can have the encoder in the encoder-decoder model that I have just mentioned try to predict a letter or a null character, and if it predicts too many null characters consecutively (we can pick a parameter for how many characters we are talking about), then we emit (on the decoder part) a 'end-of-word' token (or better say, we understand it, we don't write anywhere, this event just means that we can deliver results to the user right now, at this moment), thus we understand that a word has been pronounced and we apply the usual language modeling tricks to correct the output from the acoustic model and finally present it to the user. This approach requires a decoder that continuously listens to the output of the encoder. I am not sure if that is already implemented in flashlight, but it should be doable. To sum up: we can in theory do online speech recognition using CTC encoder-decoder model with a decoder continuously watching out for labels that get emitted from the encoder (the encoder here spits raw characters, unconstrained by any language rules, thus we might get gibberish 'words' like 'wooorrrrdd'). @tlikhomanenko  please correct me if I am wrong.

See 'Encoder-Decoder Models' here https://distill.pub/2017/ctc/

**Reply by gevorgyana** [reply]
Please note that usually encoder-decoder model is understood as in paper https://papers.nips.cc/paper/2014/file/a14ac55a4f27472c5d894ec1c3c743d2-Paper.pdf, but here in ASR, we usually have a separate component that makes sense of acoustic model predictions. This way, the decoder that I talked about is more high-level than the one that the paper talks about. In other words, you have an:
- acoustic model
- decoder (the component that makes sense of character predictions from the acoustic model)

In turn, the acoustic model can be a simple classifier (trained using CTC, e.g.), or it can be a sequence-to-sequence model (described in the paper). There are also other models available in flashlight, such as transformer-based and seq2seq with attention (main papers are mentioned in this blog post https://lilianweng.github.io/lil-log/2018/06/24/attention-attention.html, also look here to understand why transformers are special https://daleonai.com/transformers-explained)

**Reply by Grottoh** [reply]
Thanks, that's all very interesting! I expect keeping this in mind will help when looking for the relevant bits of code, and when deciding how to edit/rearange them if necessary.

Even so I fear that it will take a while to get everything right, and since it is not essential to what I'm doing right now I'm probably going to postpone this endeavor until I have more time for it; so you probably won't hear anything useful from me in the near future :(.

**Reply by gevorgyana** [reply]
I got you, okay. Still come back :3

**Reply by Grottoh** [reply]
Definitely :)

**Reply by gevorgyana** [reply]
look here https://github.com/flashlight/wav2letter/wiki/Inference-Run-Examples

**Reply by gevorgyana** [reply]
I have a proposal. We could record using gstreamer, have a small driver program for our task that would keep the beginning frame of the currently unprocessed audio sequence. Then we would try to cut a temporary file starting from that frame in the .wav that is being recorded, from the frame that we remembered till the very end. Then we would make a prediction on this sequence and see if it makes sense. If yes, then we shift the working sequence forward by remembering the right end of it as a new left end. Hope it makes sense. Maybe I will try it out.

**Reply by Grottoh** [reply]
Sounds interesting, earlier I was considering doing something like that as well. More specifically I was thinking of segmenting a stream of speech as it is being spoken into small fragments (perhaps with a bit of overlap?), encoding each fragment consecutively and then every so often decoding the emissions obtained thusfar into text (so I think more or less what you are proposing). I haven't pursued it yet because I'm new at C++ and working with audio like this did not appear trivial, but once I have time (probably in a few weeks) I will look into this as well (unless you've concluded that the idea is flawed).

By the way I had already seen the link above (https://github.com/flashlight/wav2letter/wiki/Inference-Run-Examples), however this link concerns the old pre-consolidated Wav2Letter. The code appeared to have changed quite a bit since then so I haven't looked too much at it, though indeed maybe it'll be a good idea to give it a good look at some point.

**Reply by Grottoh** [reply]
Hey, I just wanted to let you know that I have a solution that is working for me. I decided not to attempt handling (microphone) audio input using C++ since it seems unnecessarily complicated, and (for example) python has a bunch of libraries which work easily out of the box. After trying a few things I ended up with a rather simple solution which builds on the Python code in the inference notebook (https://github.com/flashlight/flashlight/blob/master/flashlight/app/asr/tutorial/notebooks/InferenceAndAlignmentCTC.ipynb).

In a nutshell, my solution is as follows:
- Start an inference process in the background (using the Python library 'subprocess')
- Then constantly record short (e.g. 1 second) audio files with your microphone
- Merge/concatenate each consecutive recording with the previous recordings, then feed it to the inference process (which translates it into text)
- After N recordings, or after the inference process detects a (relatively) long silence, the recordings that were merged so far should no longer be merged with future recordings (to prevent inference on too large audio files)

In this manner, I've been able to simply speak into my microphone and, as I'm speaking, see my speech being translated into text (in my opinion with very good accuracy), which is pretty much what I wanted. While its not entirely how I imagined it at first, and while I'm sure there are better ways of approaching this, I am happy with this for now. Though if you (or anyone else) finds a better solution then I'd of course still be curious to learn of it.

**Reply by sciai-ai** [reply]
Hi, i am trying to do the same thing. Is it possible for you to share the code. Thanks

**Reply by Grottoh** [reply]
Hey,

If you are still interested, I've just shared the code at <https://github.com/Grottoh/prototype_asr_assistant>. It has been a while since I've been busy with this project, but I'm guessing you will have to do a few things to actually get it running on your machine (e.g. download some files and organize them in the appropriate folders); I have included a few instructions which might just be enough. Much of the code probably won't interest you either, but who knows.

As I've described above, this implementation constantly records new short audio fragments which are then merged into larger ones. Though this works quite well, I do not really like having to do this. Preferably one would record once, and constantly perform inference on (copies of) this single recording, thus not having to merge recordings. Or perhaps better yet, one should make use of the streaming inference interface that MiroFurtado mentions below (though I cannot say for certain since as of yet I am unfamiliar with how this works and how well it works).

Either way, the next time I busy myself with this I think I will try to get it working using my phone somehow, so I'm unlikely to return to this code. I do not think that it is possible to install Flashlight and Wav2Letter on an android phone, but perhaps it can work in combination with a Raspberry Pi server or something, but we'll see.

Anyhow, hope that its of help to you!

**Comment by MiroFurtado** [comment, marked as answer]
I have managed to get this working with the legacy wav2letter streaming inference interface. I can try digging up some of that code at some point, this was maybe a year ago. 

Each OS has different API for microphone usage, so mine was specific to macOS. 

You might have some luck if you look under the 0.2 branch on w2l. (I believe it was 0.2?) Also follow along with some of lunixbochs issues there.

**Reply by Grottoh** [reply]
That's great, thanks! I've not been working on this project for a while now, but if I return to it I'll make sure to check it out! It does make me wonder whether this is also possible with the latest release of flashlight/wav2letter, but I think I'll also return to that question some other time.

**Comment by deepwilson** [comment]
@Grottoh Were you able to try the latest release?

**Reply by Grottoh** [reply]
Hey, I haven't looked at all this for what must be nearly two years, so I'm afraid I haven't tried it no.
